### PR TITLE
spiders: update uganda_releases with new endpoints

### LINKED ISSUE #837: `uganda_releases`: Timeout (2025-05)
https://gpppapi.com/adminapi/public/api/pdes returns 403 and `"FORTINET Webfilter This Connection is Invalid. SSL certificate expired."`

CRM-7622

### PR DESCRIPTION
**Pull request by yolile**
closes #837

### CODE REVIEW

**Comment by yolile** [comment]
@jpmckinney, requesting your review on this one as we can't get the file list from https://gpp.ppda.go.ug/public/open-data/ocds/ocds-datasets, so we need to be creative.

File: kingfisher_scrapy/spiders/uganda_releases.py [hunk 1]
```diff
@@ -1,62 +1,62 @@
...
+
+    def build_next(self, response):
+        if response.status == 500:
+            return
+        elif not self.is_http_success(response):
+            yield self.build_file_error_from_response(response)
+        else:
+            yield from super().parse(response)
+            next_code = int(parse_qs(urlparse(response.request.url).query)['code'][0])+1
+            yield self.build_request(f"{response.request.url.split('code=')[0]}code={next_code}",
```
Comment (jpmckinney) [inline kingfisher_scrapy/spiders/uganda_releases.py:55]: We can use `replace_parameters` for this line, no?

**Comment by jpmckinney** [comment]
FWIW, one of the JS files contains the data below.

However, that JS file (currently `386.461bb3e7636fac7ccee2.js`) is loaded by `runtime.[...].js`, which is loaded by the HTML page. The 386... JS file is currently the second entry in the object loaded by runtime...js, but I'm not sure that it's stable. Anyway, if their internal API changes, these facts might be helpful.

```js
                    this.ocds_data_fy = [{
                        financial_year: "2022-2023",
                        no_of_files: 4,
                        number_of_records: "3,906",
                        format: "json",
                        file_link: this.ocdsBaseFilePath + "?fy=2022-2023&code="
                    }, {
                        financial_year: "2022-2023",
                        number_of_records: "3,906",
                        format: "xlsx",
                        file_link: this.ocdsBaseFilePath + "?fy=2022-2023&code="
                    }, {
                        financial_year: "2021-2022",
                        no_of_files: 22,
                        number_of_records: "21,532",
                        format: "json",
                        file_link: this.ocdsBaseFilePath + "?fy=2021-2022&code="
                    }, {
                        financial_year: "2021-2022",
                        download_link: "",
                        number_of_records: "21532",
                        format: "xlsx"
                    }, {
                        financial_year: "2020-2021",
                        no_of_files: 24,
                        number_of_records: "23,837",
                        format: "json",
                        file_link: this.ocdsBaseFilePath + "?fy=2020-2021&code="
                    }, {
                        financial_year: "2020-2021",
                        number_of_records: "23,837",
                        format: "xlsx"
                    }, {
                        financial_year: "2019-2020",
                        no_of_files: 24,
                        number_of_records: "23,293",
                        format: "json",
                        file_link: this.ocdsBaseFilePath + "?fy=2019-2020&code="
                    }, {
                        financial_year: "2019-2020",
                        download_link: "",
                        number_of_records: "23,293",
                        format: "xlsx"
                    }]
```

File: kingfisher_scrapy/spiders/uganda_releases.py [hunk 2]
```diff
@@ -1,62 +1,62 @@
...
+    start_requests_callback = 'build_next'
+
+    def build_next(self, response):
+        if response.status == 500:
+            return
+        elif not self.is_http_success(response):
+            yield self.build_file_error_from_response(response)
+        else:
+            yield from super().parse(response)
+            next_code = int(parse_qs(urlparse(response.request.url).query)['code'][0])+1
```
[the remark this answers is not in the document]
Comment (yolile) [reply kingfisher_scrapy/spiders/uganda_releases.py:54]: Ah, right, I forgot we had these helpers, thanks!

File: kingfisher_scrapy/spiders/uganda_releases.py [hunk 3]
```diff
@@ -1,62 +1,57 @@
...
+    def build_next(self, response):
+        if response.status == 500:
+            return
+        if not self.is_http_success(response):
+            yield self.build_file_error_from_response(response)
+        else:
+            yield from super().parse(response)
+            next_code = int(get_parameter_value(response.request.url, 'code')) + 1
+            yield self.build_request(replace_parameters(response.request.url, code=next_code),
+                                     formatter=self.formatter, callback=self.build_next)
```
Comment (jpmckinney) [inline kingfisher_scrapy/spiders/uganda_releases.py:54]: Just a little simpler (build_next -> parse, next_code -> code).
```suggestion

    def parse(self, response):
        if response.status == 500:
            return
        if not self.is_http_success(response):
            yield self.build_file_error_from_response(response)
        else:
            yield from super().parse(response)
            code = int(get_parameter_value(response.request.url, 'code')) + 1
            yield self.build_request(replace_parameters(response.request.url, code=code),
                                     formatter=self.formatter)
```

**Review by jpmckinney — APPROVED** [review]

**Pull request merged**